cou/apps/core.py

```diff
@@ -221,6 +221,61 @@ def _get_disable_scheduler_step(self, units: Optional[list[Unit]]) -> list[PreUp
             for unit in units_to_disable
         ]
 
+    def _get_resume_unit_step(self, unit: Unit, dependent: bool = False) -> UnitUpgradeStep:
+        """Override the resume unit step, because extra error handling is required.
+
+        :param unit: Unit to be resumed.
+        :type unit: Unit
+        :param dependent: Whether the step is dependent of another step, defaults to False
+        :type dependent: bool, optional
+        :return: Step to resume a unit.
+        :rtype: UnitUpgradeStep
+        """
+        # workaround for https://bugs.launchpad.net/charm-ceilometer-agent/+bug/1947585
+        return UnitUpgradeStep(
+            description=(f"Resume the unit: '{unit.name}'"),
+            coro=resume_nova_compute_unit(self.model, unit),
+            dependent=dependent,
+        )
+
+
+async def resume_nova_compute_unit(model: Model, unit: Unit) -> None:
+    """Run the resume action on nova-compute, with workarounds.
+
+    Includes a workaround for https://bugs.launchpad.net/charm-ceilometer-agent/+bug/1947585
+
+    :param model: juju model to work with
+    :type model: Model
+    :param unit: nova-compute unit to resume
+    :type unit: Unit
+    :raises ActionFailed: when the resume action fails with an unknown failure
+    """
+    action = await model.run_action(unit.name, "resume", raise_on_failure=False)
+
+    # If the action was successful, there is nothing left to do
+    if action.status == "completed":
+        return
+
+    # If it failed because of https://bugs.launchpad.net/charm-ceilometer-agent/+bug/1947585 ,
+    # apply the workaround.
+    if "Services not running that should be: ceilometer-agent-compute" in action.safe_data.get(
+        "message", ""
+    ):
+        logger.debug("Resume failed because ceilometer-agent-compute not running.")
+        logger.debug("Restarting ceilometer-agent-compute on %s", unit.name)
+        await model.run_on_unit(unit.name, "sudo systemctl restart ceilometer-agent-compute")
+
+        # Update status manually, otherwise nova-compute and ceilometer-agent
+        # will be blocked until next update-status hook.
+        await model.update_status(unit.name)
+        for subordinate in unit.subordinates:
+            if subordinate.charm == "ceilometer-agent":
+                await model.update_status(subordinate.name)
```

Comment:
After this, everything should be active/idle again.  I don't think we need to re-run the `resume` action?

Reply:
should be